shell/platform/windows/keyboard_key_embedder_handler.cc

```diff
@@ -0,0 +1,406 @@
+                         kFlutterKeyEventTypeUp, key_info.physical_key,
+                         key_info.logical_key, empty_character),
+                     nullptr, nullptr);
+        } else {
+          // This key will always be pressed in the following synthesized event.
+          pressingRecords_[key_info.physical_key] = key_info.logical_key;
+        }
+        sendEvent_(SynthesizeSimpleEvent(kFlutterKeyEventTypeDown,
+                                         key_info.physical_key,
+                                         key_info.logical_key, empty_character),
+                   nullptr, nullptr);
+      }
+      key_info.toggled_on = should_toggled;
+    }
+  }
+}
+
+void KeyboardKeyEmbedderHandler::SynchroizeCritialPressedStates() {
+  for (auto& kv : critical_keys_) {
+    UINT virtual_key = kv.first;
+    CriticalKey& key_info = kv.second;
+    if (key_info.physical_key == 0) {
+      // Never seen this key.
+      continue;
+    }
+    assert(key_info.logical_key != 0);
+    SHORT state = get_key_state_(virtual_key);
+    if (key_info.check_pressed) {
+      auto recorded_pressed_iter = pressingRecords_.find(key_info.physical_key);
+      bool recorded_pressed = recorded_pressed_iter != pressingRecords_.end();
+      bool should_pressed = state & kStateMaskPressed;
+      if (recorded_pressed != should_pressed) {
+        if (should_pressed) {
+          pressingRecords_[key_info.physical_key] = key_info.logical_key;
+        } else {
+          pressingRecords_.erase(recorded_pressed_iter);
+        }
+        const char* empty_character = "";
+        sendEvent_(
+            SynthesizeSimpleEvent(should_pressed ? kFlutterKeyEventTypeDown
+                                                 : kFlutterKeyEventTypeUp,
+                                  key_info.physical_key, key_info.logical_key,
+                                  empty_character),
+            nullptr, nullptr);
+      }
+    }
+  }
+}
+
+void KeyboardKeyEmbedderHandler::HandleResponse(bool handled, void* user_data) {
+  PendingResponse* pending = reinterpret_cast<PendingResponse*>(user_data);
+  auto callback = std::move(pending->callback);
+  callback(handled, pending->response_id);
+}
+
+void KeyboardKeyEmbedderHandler::InitCriticalKeys() {
+  auto createCheckedKey = [this](UINT virtual_key, bool extended,
+                                 bool check_pressed,
+                                 bool check_toggled) -> CriticalKey {
+    UINT scan_code = MapVirtualKey(virtual_key, MAPVK_VK_TO_VSC);
+    return CriticalKey{
+        .physical_key = getPhysicalKey(scan_code, extended),
+        .logical_key = getLogicalKey(virtual_key, extended, scan_code),
+        .check_pressed = check_pressed || check_toggled,
+        .check_toggled = check_toggled,
+        .toggled_on = check_toggled
+                          ? !!(get_key_state_(virtual_key) & kStateMaskToggled)
+                          : false,
+    };
+  };
+
+  critical_keys_.emplace(VK_LSHIFT,
```

Comment:
Are there any others that need consideration here?  I'm thinking of things like AltGr, various lock keys for Japanese input modes, shift-lock (as opposed to caps lock), etc.

Also, I wonder how the "sticky keys" accessibility setting affects these key events: probably they are simulated at the system level and we won't need to worry about it, but it wouldn't hurt to try it out.

Reply:
You're right, although I'd prefer patching it afterwards with an actual hardware. The specification does not always match from my manual test result, namely left and right `Alt` are not distinguished by `LMENU` and `RMENU` but the extended bits.

Reply:
Let me add a TODO here.